File: iosApp/iosApp/service/AccountService.swift
//
// Created by Tim Ortel on 26.09.22.
//

import Foundation
import Combine

/**
 * Service that provides data about the users login status.
 */
protocol AccountService {

    /**
     * The latest authentication data. Publisher emits a new element whenever the login status of the user changes.
     */
    var authenticationData: any Publisher<AuthenticationData, Never> { get }

    func login(username: String, password: String, rememberMe: Bool) async -> LoginResponse

    /**
     - Returns: If the user is currently logged in.
     */
    func isLoggedIn() -> Bool

    /**
     * Deletes the JWT.
     */
    func logout()
}

/**
 * Can either be [LoggedIn] or [NotLoggedIn].
 */
enum AuthenticationData {
    case NotLoggedIn
    case LoggedIn(authToken: String)
}

struct LoginResponse {
    let isSuccessful: Bool
}

File: appCommon/src/commonMain/kotlin/de/tum/informatics/www1/artemis/native_app/service/AccountService.kt
package de.tum.informatics.www1.artemis.native_app.service

import kotlinx.coroutines.flow.Flow

/**
 * Service that provides data about the users login status.
 */
interface AccountService {

    /**
     * The latest authentication data. Flow emits a new element whenever the login status of the user changes.
     */
    val authenticationData: Flow<AuthenticationData>

    /**
     * Can either be [LoggedIn] or [NotLoggedIn].
     */
    sealed class AuthenticationData {
        object NotLoggedIn : AuthenticationData()

        data class LoggedIn(val authToken: String) : AuthenticationData() {
            val asBearer = "Bearer $authToken"
        }
    }

    suspend fun login(username: String, password: String, rememberMe: Boolean): LoginResponse

    data class LoginResponse(val isSuccessful: Boolean)

    /**
     * Deletes the JWT.
     */
    suspend fun logout()
}